### PR TITLE
feat: New resource for standalone project users and groups

### PR DESCRIPTION
**Pull request by Danielku15**
This PR superceeds #99 with the capabilitiy to both add users and groups individually to projects. 

This way we cover both issues #98 and #83 in one go. 

## project_user
It adds a new dedicated resource which allows managing project members without touching the project itself. In our Company-wide Artifactory instance we are given 1 project by the admins and the project is mostly readonly. We can add members and manage repositories but nothing else. 

With this resource we can:

1. Add new project members to the existing project without touching it. 
2. We can still manage some memberships manually without interference with the terraform state.
3. We can auto-generate a list of users who should have theoretically access IF the accounts exist and still apply without errors (only warnings). This serves the use case where we have a central list of developers and we don't know if they already have an account in artifactory (e.g. missing SSO login). We will see who is missing a user without failing to apply users who have one. 

Unfortunately I cannot run the tests on our company artifactory instance due to permissions. But I tried to prepare already some tests. Any help in finalizing the tests is appreciated (providing an artifactory for testing or taking over to write the tests). 

I made some manual tests by using the locally compiled provider and creating members in our project. 

## project_group

Similar like project_user just simplified settings (no ignore missing feature needed here). I couldn't do proper tests in our artifactory instance as we cannot manage groups and no test data is suitable.

### CODE REVIEW

File: pkg/project/resource_project_group_test.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+			]
+		}
+
+		resource "project_group" "{{ .group }}" {
+			project_key = "{{ .project_key }}"
+			name = "{{ .group }}"
+			roles = {{ .roles }}
+
+			depends_on = [
+				project.{{ .project_name }}
```
Comment (alexhung) [inline pkg/project/resource_project_group_test.go:61]: This is not necessary if you have `project_key = project.{{ .project_name }}.key` as Terraform will figure out the dependency tree.

File: pkg/project/resource_project_group_test.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+			}
+
+			depends_on = [
+				artifactory_group.{{ .group }}
+			]
+		}
+
+		resource "project_group" "{{ .group }}" {
+			project_key = "{{ .project_key }}"
+			name = "{{ .group }}"
```
Comment (alexhung) [inline pkg/project/resource_project_group_test.go:57]: If you use `name = artifactory_group.{{ .group }}.name` then you won't need the `depends_on` in `project` resource.

File: pkg/project/resource_project_group_test.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+	"github.com/go-resty/resty/v2"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/jfrog/terraform-provider-shared/test"
+)
+
+func TestAccProjectGroup(t *testing.T) {
+	projectName := fmt.Sprintf("tftestprojects%s", randSeq(10))
+	projectKey := strings.ToLower(randSeq(6))
+
+	group := "group1"
```
Comment (alexhung) [inline pkg/project/resource_project_group_test.go:17]: We should randomize the group name

File: pkg/project/resource_project_group_test.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+				VersionConstraint: "10.1.3",
+			},
+		},
+		Steps: []resource.TestStep{
+			{
+				Config: config,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "project_key", fmt.Sprintf("%s", params["project_key"])),
+					resource.TestCheckResourceAttr(resourceName, "name", group),
+					resource.TestCheckResourceAttr(resourceName, "ignore_missing_user", "false"),
```
Comment (alexhung) [inline pkg/project/resource_project_group_test.go:95]: This is invalid test check. Group doesn't have this attribute 😄

File: pkg/project/resource_project_user_test.go [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+	"github.com/go-resty/resty/v2"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/jfrog/terraform-provider-shared/test"
+)
+
+func TestAccProjectUser(t *testing.T) {
+	projectName := fmt.Sprintf("tftestprojects%s", randSeq(10))
+	projectKey := strings.ToLower(randSeq(6))
+
+	username := "user1"
```
Comment (alexhung) [inline pkg/project/resource_project_user_test.go:18]: Let's randomize the user name.

**Review by alexhung — CHANGES_REQUESTED** [review]
Look good overall. I have a few comments on test configurations dependencies, and test checks. 

We also need to bump the `project` schema version and add new attribute similar to `roles` to ease migration to use the new `project_user` and project_group` attributes: https://github.com/jfrog/terraform-provider-project/blob/master/pkg/project/resource_project.go#L266 At the very least we need deprecation messages to warn users of the new attributes.

**Comment by Danielku15** [comment]
Thanks for the remarks. I'm still quite new to developing Go and terraform providers 😁 

If there is some good way for me to run the test suite on my side I'm happy to do so. It's a bit unfortunate that there is no easy to use test environment due to license needs. 

Would be nice if there would be an Artifactory docker image which works for a limited time before it shuts down for such testing purposes. I'm thinking of a built-in 1hour trial with full data-loss (to prevent misuse). 😆

**Comment by alexhung** [comment]
If there is some good way for me to run the test suite on my side I'm happy to do so. It's a bit unfortunate that there is no easy to use test environment due to license needs.

@Danielku15 A contributor in https://github.com/jfrog/vault-plugin-secrets-artifactory uses the JFrog JCR Docker image to test against. It doesn't require license but I'm pretty sure there are feature limitations.

Case c) is what I'd like to see. Setting `Deprecate` string **does not** actually remove the attribute or change the state. It simply makes Terraform CLI to emit that message during use to warn users about the deprecation. If you don't feel comfortable setting it, I will after this PR is merged 😄

**Review by alexhung — APPROVED** [review]
LGTM

**Review by danielmkn — APPROVED** [review]

**Review by alexhung — APPROVED** [review]
Looks good. I can take over from here.

**Comment by alexhung** [comment]
> JFrog might consider making rather 1 provider per "API" with 1:1 mapping for resources as things are a bit inconsistent.

@Danielku15 We are working on improving the Projects API so the dependency on Artifactory API will be removed.

**Pull request merged**